### ISSUE TITLE
View updates may cause reading and deserializing unused columns

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#10062]
When view updates are generated, we create base table readers to read rows if read-before-write is used. It's often the case that we don't need all columns from the row, but a subset of them - specifically, a set union of all columns selected by all views that we're going to generate the updates for. When the base row size has the same order of magnitude as the view row this is more or less harmless, but if the base row is considerably larger - e.g. it has a huge collection, not selected by any views - we'll pay a hefty price for that.

Namely, we'll not only have large redundant read amplification for reading the unused columns, but also we'll deserialize them, which is a significant CPU & memory effort for collections. What's more, we'll also deserialize everything from the incoming mutation (from which we generate the view update) and that includes deserializing the huge collection too, even though we never use it later.

This is a side-effect of generating view updates with row granularity - we simply operate on full rows, because it's easier, but it can also produce huge overheads, as (most likely) observed in #10044.

To sum up, we should strive to avoid:
1. deserializing unused columns from the base table mutation when generating the view updates - it's wasted CPU and memory, especially for large collections
2. reading unused columns from disk for read-before-write - it's wasted I/O

/cc @slivne @nyh

**Comment by psarna** [comment]
This is one of the offending lines in terms of redundant deserialization:
https://github.com/scylladb/scylla/blob/00a6fda7b98438184024fc4683b0accf1852e30c/db/view/view.cc#L709

It produces a difference of the updated entry and the previous entry, but it disregards the fact that we only care about certain columns, so it happily deserializes columns unselected by the view as well.

Still, CPU and memory overhead is one thing, but we should start by not reading the whole row from disk in the first place, if we only need specific columns (unless we always operate with row granularity anyway, when it comes to reading from sstables?).

**Comment by slivne** [comment]
We will need to read the complete row - if we want to place the data in the
cache  - the cache works in row level not in cell level.

**Comment by nyh** [comment]
> Sigh, it all might be on purpose, just to make sure that the liveness intricacies work... as well as potential filtering (in other words, reading a 10MiB collection in order to get its ttl):

These "liveness intricacies" only apply if **the view key does not have a new key column beyond the key of the base**. In that case (see issue #3362 why we have this) we indeed need to copy even unselected columns as "virtual columns" (a column with a timestamp but empty value). For a large collection we indeed need to copy all its keys with timestamps and empty values, so we need to read - and write - the whole thing. 
When the view does have a new key column, we don't need those "virtual columns" so this problem doesn't exist, and you're right, we could perhaps save some time on deserializing the large collection (although Avi is right, it's not clear if we have the API to do it).

However, as I noted in #3594, even when we do need to write virtual columns for _unselected_ columns so in theory need to read everything including unselected columns, not all writes really need to do read-before-write at all: In some cases not only do we not need to read the entire row, we don't need to read anything at all: If the base update did not view's _key_, and only modified regular columns of the view, then we don't need read-before write at all. We implemented a hack in #3819 which reduces writes in this case, but I think it doesn't avoid the read.

One of the problems in this area is that it's not easy to keep track of what the code really does and does it do exactly what you think it does in every one of the cases mentioned above (and many more). I think we need to write test cases which build various MV and performs reads and checks metrics (hoping we have enough metrics for that) to see whether we read too many bytes - and of course also checks correctness of the view.
